core/src/main/java/org/pentaho/platform/util/Emailer.java

```diff
@@ -261,89 +289,105 @@ public boolean send() {
-            } else {
-              //CID attachments
-              newMultipart.addBodyPart( bodyPart );
-            }
-          }
-
+          setBodyForEmbeddedHtml( msg, body );
 
-          msg.setContent( newMultipart );
         }
       } else {
 
         // construct the message
         msg = new MimeMessage( session );
-        Multipart multipart = new MimeMultipart();
-
-        if ( body != null ) {
-          MimeBodyPart bodyMessagePart = new MimeBodyPart();
-          bodyMessagePart.setText( body, LocaleHelper.getSystemEncoding() );
-          multipart.addBodyPart( bodyMessagePart );
-        }
-
-        if ( attachment != null ) {
-          ByteArrayDataSource dataSource = new ByteArrayDataSource( attachment, attachmentMimeType );
-          // attach the file to the message
-          MimeBodyPart attachmentBodyPart = new MimeBodyPart();
-          attachmentBodyPart.setDataHandler( new DataHandler( dataSource ) );
-          attachmentBodyPart.setFileName( attachmentName );
-          multipart.addBodyPart( attachmentBodyPart );
-        }
+        Multipart multipart = setMultipart( body );
 
         // add the Multipart to the message
         msg.setContent( multipart );
       }
 
-      if ( from != null ) {
-        msg.setFrom( new InternetAddress( from, fromName ) );
-      } else {
-        // There should be no way to get here
-        logger.error( "Email.ERROR_0012_FROM_NOT_DEFINED" ); //$NON-NLS-1$
-      }
-
-      if ( ( to != null ) && ( to.trim().length() > 0 ) ) {
-        msg.setRecipients( Message.RecipientType.TO, InternetAddress.parse( to, false ) );
-      }
-      if ( ( cc != null ) && ( cc.trim().length() > 0 ) ) {
-        msg.setRecipients( Message.RecipientType.CC, InternetAddress.parse( cc, false ) );
-      }
-      if ( ( bcc != null ) && ( bcc.trim().length() > 0 ) ) {
-        msg.setRecipients( Message.RecipientType.BCC, InternetAddress.parse( bcc, false ) );
-      }
-
-      if ( subject != null ) {
-        msg.setSubject( subject, LocaleHelper.getSystemEncoding() );
-      }
-
-
+      setToFromSubject( from, fromName, to, cc, bcc, subject, msg );
       msg.setHeader( "X-Mailer", Emailer.MAILER ); //$NON-NLS-1$
       msg.setSentDate( new Date() );
 
-      Transport.send( msg );
+      if ( service != null ) {
+        service.sendEmail( emailConfig, token, session, msg );
+      } else {
+        Transport.send( msg );
+      }
 
       return true;
     } catch ( SendFailedException e ) {
       logger.error( "Email.ERROR_0011_SEND_FAILED -" + to, e ); //$NON-NLS-1$
     } catch ( AuthenticationFailedException e ) {
       logger.error( "Email.ERROR_0014_AUTHENTICATION_FAILED - " + to, e ); //$NON-NLS-1$
-    } catch ( Throwable e ) {
+    } catch ( Exception e ) {
       logger.error( "Email.ERROR_0011_SEND_FAILED - " + to, e ); //$NON-NLS-1$
     }
     return false;
   }
+
+  private Session getSmtpSession( boolean authenticate ) {
```

Comment:
misleading name this function is used for Oauth as well probably just `getSession`